File: 2846-robot-collisions/robot-collisions.java
import java.util.*;

class Solution {
    public List<Integer> survivedRobotsHealths(int[] positions, int[] healths, String directions) {
        int n = positions.length;
        Integer[] indices = new Integer[n];
        
        // Initialize indices 0 to n-1
        for (int i = 0; i < n; i++) {
            indices[i] = i;
        }

        // Sort indices based on the robots' positions on the line
        Arrays.sort(indices, (i, j) -> Integer.compare(positions[i], positions[j]));

        Stack<Integer> stack = new Stack<>();

        // Process robots from left to right
        for (int curr : indices) {
            // If moving right, it might collide with future 'L' robots, so add to stack
            if (directions.charAt(curr) == 'R') {
                stack.push(curr);
            } else {
                // If moving left, resolve collisions with any 'R' robots in the stack
                while (!stack.isEmpty() && directions.charAt(stack.peek()) == 'R' && healths[curr] > 0) {
                    int top = stack.peek();
                    
                    if (healths[curr] > healths[top]) {
                        healths[curr] -= 1;   // Current robot takes 1 damage
                        healths[top] = 0;     // Top robot is destroyed
                        stack.pop();
                    } else if (healths[curr] < healths[top]) {
                        healths[top] -= 1;    // Top robot takes 1 damage
                        healths[curr] = 0;    // Current robot is destroyed
                    } else {
                        // Both have equal health and destroy each other
                        healths[curr] = 0;
                        healths[top] = 0;
                        stack.pop();
                    }
                }
                
                // If the 'L' robot survived all collisions, it continues moving left harmlessly
                if (healths[curr] > 0) {
                    stack.push(curr);
                }
            }
        }

        // Collect the surviving robots' healths in their original given order
        List<Integer> result = new ArrayList<>();
        for (int h : healths) {
            if (h > 0) {
                result.add(h);
            }
        }
        
        return result;
    }
}

File: 2846-robot-collisions/robot-collisions.cpp
#include <vector>
#include <string>
#include <numeric>
#include <algorithm>

using namespace std;

class Solution {
public:
    vector<int> survivedRobotsHealths(vector<int>& positions, vector<int>& healths, string directions) {
        int n = positions.size();
        
        // Array to keep track of original indices
        vector<int> ids(n);
        iota(ids.begin(), ids.end(), 0); 
        
        // Sort indices based on starting positions from left to right
        sort(ids.begin(), ids.end(), [&](int i, int j) {
            return positions[i] < positions[j];
        });

        vector<int> st; // Stack to hold indices of processed robots
        
        for (int id : ids) {
            if (directions[id] == 'R') {
                // Moving right, no immediate collision, push to stack
                st.push_back(id);
            } else {
                // Moving left, process potential collisions
                while (!st.empty() && directions[st.back()] == 'R' && healths[id] > 0) {
                    int top_id = st.back();
                    
                    if (healths[id] > healths[top_id]) {
                        // Current 'L' robot wins
                        healths[id] -= 1;
                        healths[top_id] = 0; // Top robot is destroyed
                        st.pop_back();
                    } else if (healths[id] < healths[top_id]) {
                        // Top 'R' robot wins
                        healths[top_id] -= 1;
                        healths[id] = 0; // Current robot is destroyed
                    } else {
                        // Tie: both are destroyed
                        healths[id] = 0;
                        healths[top_id] = 0;
                        st.pop_back();
                    }
                }
                
                // If the current 'L' robot survived all collisions, it joins the survivors
                if (healths[id] > 0) {
                    st.push_back(id);
                }
            }
        }
        
        // Collect surviving robots in their original order
        vector<int> result;
        for (int i = 0; i < n; ++i) {
            if (healths[i] > 0) {
                result.push_back(healths[i]);
            }
        }
        
        return result;
    }
};